rslib/src/notetype/mod.rs

```diff
@@ -272,6 +272,103 @@ impl Notetype {
             });
     }
 
+    fn ensure_template_fronts_unique(&self) -> Result<()> {
+        let mut map = HashMap::new();
+        if let Some((index_1, index_2)) =
+            self.templates.iter().enumerate().find_map(|(index, card)| {
+                map.insert(&card.config.q_format, index)
+                    .map(|old_index| (old_index, index))
+            })
+        {
+            Err(AnkiError::TemplateSaveError(TemplateSaveError {
+                notetype: self.name.clone(),
+                ordinal: index_2,
+                details: TemplateSaveErrorDetails::Duplicate(index_1),
+            }))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Ensure no templates are None, every front template contains at least one
+    /// field, and all used field names belong to a field of this notetype.
+    fn ensure_valid_parsed_templates(
+        &self,
+        templates: &[(Option<ParsedTemplate>, Option<ParsedTemplate>)],
+    ) -> Result<()> {
+        if let Some((invalid_index, details)) =
+            templates.iter().enumerate().find_map(|(index, sides)| {
+                if let (Some(q), Some(a)) = sides {
+                    let q_fields = q.fields();
+                    if q_fields.is_empty() {
+                        Some((index, TemplateSaveErrorDetails::NoFrontField))
+                    } else if self.unknown_field_name(q_fields.union(&a.fields())) {
+                        Some((index, TemplateSaveErrorDetails::NoSuchField))
+                    } else {
+                        None
+                    }
+                } else {
+                    Some((index, TemplateSaveErrorDetails::TemplateError))
+                }
+            })
+        {
+            Err(AnkiError::TemplateSaveError(TemplateSaveError {
+                notetype: self.name.clone(),
+                ordinal: invalid_index,
+                details,
+            }))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// True if any name in names does not belong to a field of this notetype.
+    fn unknown_field_name<T, I>(&self, names: T) -> bool
+    where
+        T: IntoIterator<Item = I>,
+        I: AsRef<str>,
+    {
+        let special_fields = [
```

Comment:
The only other place I could find where we list out special fields is `notetype/render.rs/add_special_fields()`. It would probably make sense to exhaustively iterate over a list of all special fields like this array here.
Should I maybe move the array into `render.rs` and make it public?

Reply:
If we'll use it in multiple locations, by all means make it pub(crate). If you meant completely public, I'd prefer to keep things limited to the crate unless there's a clear use case for outside consumers, as it's extra API surface we need to support

Reply:
I can't get it to work, unfortunately. If I do it like this:

```rust
SPECIAL_FIELDS.iter().map(|&field| match field {
            "Tags" => map.entry(field).or_insert_with( || tags.into()),
            "Type" => map.entry(field).or_insert_with(|| nt.name.clone().into()),
            "Subdeck" => map.entry(field).or_insert_with(|| subdeck_name.to_string().into()),
            "Deck" => map.entry(field).or_insert_with(|| deck_name.to_string().into()),
            "CardFlag" => map.entry(field).or_insert_with(|| flag_name(card.flags).into()),
            "Card" => map.entry(field).or_insert_with(|| template.name.clone().into()),
            // handled by Python
            "FrontSide" => "",
        });
```

I get ```captured variable cannot escape `FnMut` closure body```, which makes sense because the compiler can't know that every closure will only be called once.
Do you know of an idiomatic way to ensure we have handled (almost) all values in the array?
I could write a test, but this wouldn't work at compile time, of course.

Reply:
You could perhaps accomplish it with strum, but if it proves too complicated, feel free to duplicate the list and just leave comments pointing to the other location that needs to be changed when updates are made - it's not an area that gets updated frequently.